### ISSUE TITLE
Simp regressions probably related to transparency

### ISSUE THREAD

**Issue by PatrickMassot** [leanprover/lean4#2670]
### Prerequisites

* [C] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

Many `simp` proofs from Lean 3 no longer work, probably because the discrimination tree that is necessary for improved performance does not allow seeing through definitional equality. I think this is worth reporting anyway.

### Context

The specific examples come from [this Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/simp.20.5BX.5D.20fails.2C.20rw.20.5BX.5D.20works/near/390538357)

### Steps to Reproduce

```lean
example {α β : Type} {f : α × β → β → β} (h : ∀ p : α × β, f p p.2 = p.2)
  (a : α) (b : β) : f (a, b) b = b := by
  -- simp [h] -- fails but worked in Lean 3
  rw [h] -- works

example {α β : Type} {f : α × β → β → β}
  (a : α) (b : β) (h : f (a,b) (a,b).2 = (a,b).2) : f (a, b) b = b := by
  -- simp [h] -- fails but worked in Lean 3
  rw [h] -- works
```

**Expected behavior:** 

The `simp` call closes the goal as they did in Lean 3.

**Actual behavior:** 

`simp` is not working but `rw` works.

### Versions

Lean (version 4.2.0-rc1, commit https://github.com/leanprover/lean4/commit/a62d2fd4979671b76b8ab13ccbe4fdf410ec0d9d, Release)
Ubuntu 22.04

### Impact

In isolation this is not a problem, but when porting project it can make a complicated simp call involving a lot of lemmas very painful to debug. And the ported output then requires alternating calls to `simp only` and `rw`.

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by leodemoura** [comment]
Yes, it is due to discriminations trees. We can disable the discrimination tree indexing for a subterm `t` using `no_index t`. Example:
```lean
example {α β : Type} {f : α × β → β → β} (h : ∀ p : α × β, f p (no_index p.2) = p.2)
  (a : α) (b : β) : f (a, b) b = b := by
  simp [h]


example {α β : Type} {f : α × β → β → β}
  (a : α) (b : β) (h : f (a,b) (no_index (a,b).2) = (a,b).2) : f (a, b) b = b := by
  simp [h]
```

A possible workaround: global `@[simp]` theorems do not usually have LHS's with reducible terms such as `(a, b).2`, but this kind of term does occur in local hypotheses. One option is to instruct Lean to always use `no_index` for the arguments of local hypotheses. We would recover the Lean 3 behavior for them. Drawback: performance impact on tactics such as `simp_all` that may operate in big local contexts.

**Comment by digama0** [comment]
> global @[simp] theorems do not usually have LHS's with reducible terms such as (a, b).2,

Unfortunately, these do exist in the category theory library, usually deep inside some implicit type arguments in dependencies of the main arguments (which makes manually adding `no_index` quite painful in practice). Possibly this could be addressed by preprocessing the LHS of simp theorems when they are added to the tree to dsimp them away, although this is problematic when the lemma is itself about this reduction (i.e. `@[simp] theorem mk_snd : (a, b).2 = b`).

**Comment by leodemoura** [comment]
@digama0 Could you please provide evidence for your claim? It is not clear to me why someone will write `(a, b).2` in a theorem statement if they want to be able to match with `b`.

**Comment by digama0** [comment]
They don't write it themselves, it is generated by elaboration up in the implicit type dependencies that users don't see. This is a half-remembered example from when I had to debug such an issue, the text `(a, b).2` never appears in the source text and you only notice it appearing in the elaboration trace if you dig into the terms. I will see if I can locate the example.

**Comment by digama0** [comment]
@leodemoura Here's [the issue](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/simp.20.5BX.5D.20fails.2C.20rw.20.5BX.5D.20works/near/359872281) I was recalling. As you can see from the [debugging discussion](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/simp.20.5BX.5D.20fails.2C.20rw.20.5BX.5D.20works/near/359905996), the term `(W, X).fst` shows up as the inferred argument to a function with some complex dependencies, even though it was never explicitly written down.

**Comment by leodemoura** [comment]
@digama0 Could you please show me the actual Mathlib theorem statement that contains `(W, X).fst`? Moreover, if it is an implicit argument, it should not matter since the discrimination tree ignores them.

**Comment by PatrickMassot** [comment]
I'm sorry about the confusion brought by my second example. The first failure was the one I actually encountered when porting the sphere eversion project. The second one was brought up during the Zulip discussion. I just checked that `no_index` [works in the example that triggered the discussion](https://github.com/leanprover-community/sphere-eversion/commit/5c8f6b56738be05b8b8827ba98165c706c8a5e8a).
